File: rust/2021/src/bin/day_14.rs
use aoc::counter::Counter;
use std::collections::HashMap;
use std::fs;

fn simulate(rule_map: &HashMap<(u8, u8), u8>, bytes: &[u8], steps: usize) -> usize {
    // Track the number of elements and pairs
    let mut elements: Counter<_> = bytes.iter().copied().collect();
    let mut pairs: Counter<_> = bytes
        .iter()
        .copied()
        .zip(bytes.iter().copied().skip(1))
        .collect();
    // Run simulation
    for _ in 0..steps {
        let prev = pairs;
        pairs = Counter::new();
        for (p @ (a, b), n) in prev {
            let c = rule_map[&p];
            elements[&c] += n;
            for pair in [(a, c), (c, b)] {
                pairs[&pair] += n;
            }
        }
    }
    let least = elements.values().min().unwrap();
    let most = elements.values().max().unwrap();
    most - least
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_14.txt").unwrap();
    let (template, rules) = input.split_once("\n\n").unwrap();
    let template_bytes = template.as_bytes();
    // Map element pairs to elements
    let rule_map: HashMap<_, _> = rules
        .lines()
        .map(|line| {
            let (pair, elem) = line.split_once(" -> ").unwrap();
            let bytes = pair.as_bytes();
            ((bytes[0], bytes[1]), elem.as_bytes()[0])
        })
        .collect();
    println!("Part 1: {}", simulate(&rule_map, template_bytes, 10));
    println!("Part 2: {}", simulate(&rule_map, template_bytes, 40));
}

File: haskell/2021/day_14.hs
import qualified Control.Monad as Monad
import qualified Data.Char as Char
import qualified Data.List.Split as Split
import qualified Data.Map as Map
import qualified Data.Set as Set

type Pair = (Char, Char)

-- | Perform one step of the simulation.
step :: Map.Map Pair Char -> Map.Map Pair Int -> Map.Map Pair Int
step ruleMap pairs = Map.fromListWith (+) $ do
    (pair@(a, b), n) <- Map.assocs pairs
    let c = ruleMap Map.! pair
    [((a, c), n), ((c, b), n)]

-- | Perform polymer expansion and return the difference between
-- most and least common elements.
simulate:: Map.Map Pair Char -> String -> Int -> Int
simulate ruleMap template steps =
    let pairs = zip template $ tail template
        initial = Map.fromListWith (+) $ zip pairs (repeat 1)
        final = iterate (step ruleMap) initial !! steps
        -- Count elements using pair counts
        elements = Map.fromListWith (+) $ do
            ((a, b), n) <- Map.assocs final
            [(a, n), (b, n)]
        -- Divide counts by 2; add 1 for the start and end
        fixed = Map.mapWithKey (\k v ->
            let extra = length $ filter (==k) [head template, last template]
            in v `div` 2 + extra
            ) elements
    in maximum (Map.elems fixed) - minimum (Map.elems fixed)

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_14.txt"
    let template:_:rules = lines text
    let ruleMap = Map.fromList [((a, b), c) | a:b:_:_:_:_:c:_ <- rules]
    putStrLn $ "Part 1: " ++ show (simulate ruleMap template 10)
    putStrLn $ "Part 2: " ++ show (simulate ruleMap template 40)
